examples/metrics/grafana-dashboards/strimzi-kafka.json

```diff
@@ -1027,7 +1027,7 @@
       "steppedLine": false,
       "targets": [
         {
-          "expr": "sum(jvm_memory_bytes_used{namespace=\"$kubernetes_namespace\",kubernetes_pod_name=~\"$strimzi_cluster_name-$kafka_broker\",strimzi_io_name=\"$strimzi_cluster_name-kafka\"}) by (kubernetes_pod_name)",
+          "expr": "sum(jvm_memory_bytes_used{namespace=\"$kubernetes_namespace\",kubernetes_pod_name=~\"$strimzi_cluster_name-$kafka_broker\",strimzi_io_name=\"$strimzi_cluster_name-kafka-bootstrap\"}) by (kubernetes_pod_name)",
```

Comment:
I think this is confusing - the metrics do not come from the service but from the pod. Is there any reason we don't use PodMonitors to scrape all pods? Why do we use ServiceMonitor somewhere and PodMonitor elsewhere?

Reply:
The ServiceMonitor is translated into `endpoints` role discovery which actually really scrapes from the backed pods.
About PodMonitor vs ServiceMonitor you can read more here [2]

[1] https://prometheus.io/docs/prometheus/latest/configuration/configuration/#endpoints
[2] https://github.com/coreos/prometheus-operator/issues/3119

> In a sense this also impacts backwards compatibility.

What kind of backwards compatibility are you talking about? We were using the `kubernetes-services` which was a scraping job with `endpoints` role as well.

Reply:
> What kind of backwards compatibility are you talking about? We were using the kubernetes-services which was a scraping job with endpoints role as well.

Well, you are clearly changing the labels. So users will need to update the Prometheus settings to be able to use the new dashboards. Right? Not saying this is a blocker, but it is for sure not user friendly and might be worth pointing out in change log. But keeping the old labels would for sure be better as it creates less disruption.

> The ServiceMonitor is translated into endpoints role discovery which actually really scrapes from the backed pods.

I know that, but if the ServiceMonitor means changes to labels, I would see PodMonitor as better option if it keeps the labels the same as we had before this change. I do not think the links list any actual reasons why one would be preferred over the other.

Reply:
The usage of the new dashboards come with the usage with the new Prometheus configuration, users should not apply new dashboards leaving ServiceMonitor and additional configuration as they are today, for sure it would brake compatibility.

Regarding ServiceMonitor, it's clear to me that it came before PodMonitor and that the last one was added only for scraping pods that are not behind a service. People should prefer ServiceMonitor when you have service as in our case. At the end it doesn't scrape the service actually, but the endpoints behind it.

See this as well https://github.com/coreos/prometheus-operator/issues/38

Reply:
> Regarding ServiceMonitor, it's clear to me that it came before PodMonitor and that the last one was added only for scraping pods that are not behind a service. People should prefer ServiceMonitor when you have service as in our case. At the end it doesn't scrape the service actually, but the endpoints behind it.

But why should people prefer ServiceMonitor? It is IMHO important - should we create services and use service monitors where we now use PodMonitors? If it doesn't have any negative consequences and we are fine to use PodMonitors there, maybe we can use them here as well instead of changing the labels and breaking the compatibility?

Reply:
Tbh I think it makes sense to use PodMonitor for things like CO and EO because there are actually no services backed by them.
I was wondering that maybe the bridge as well should be moved to use ServiceMonitor instead of PodMonitor.
And if it's all about backward compatibility I can try to relable in the way to not change the dashboard ... but I don't agree tbh. I don't see any problem. I think that a dashboard update is really tight to a Prometheus configuration update if needed.

Reply:
Ok. Then please add docs for the upgrade of the Prometheus configuration and the dashboards. We had issues in the past with this, so we need to start doing it better. We should either in the docs or in the CHANGELOG cover what changed for users using their own Prometheus deployments + explain the changes to the users using the deployment we provide (what to delete, what to create etc.).

Reply:
When you say "their own Prometheus deployments" what do you mean exactly? 
I agree that users can create their Prometheus instance in the way they want but if they are using our stuff then it means they are using the Prometheus operator at least translating PodMonitor and ServiceMonitor to "pod" and "endpoints" configs + additional config we provide. These files already provide some relabeling like `kubernetes_pod_name` so if they are using our dashboards they should anyway use our PodMonitor, ServiceMonitor, etc etc. This is the reason why I see dashboard and scraping jobs really tight.

Reply:
I think there is from the beginning an assumption that users do not have to deploy _Strimzi Prometheus_ but just plugin Strimzi into their own Prometheus whatever way they deploy it. So I think we need to have a not somewhere saying that from 0.19.0, the Grafana dashboards expect the metrics to be scraped with the service labels and not pod labels as it was before.

Reply:
I mean you cannot expect the users to have one Prometheus deployed for Strimzi this way and another Prometheus deployed for some SQL database with some other tooling they use etc. That does not make sense.

Reply:
I agree but plugging our metrics into their Prometheus instance actually mean deploy our PodMonitor and our ServiceMonitor (this is also why the additional configuration should go away and that is only about node related stuff). So one more time this is why I see PM and SM related to dashboards; you need the ones we provided to scrape Strimzi related metrics.

Reply:
But they are coupled to a particular operator when you create that as a CR. There might be users configuring it all in a single Prometheus config map with their own deployment not based on any operators for example.

Reply:
I see your point now, for people not using the Prometheus operator to configure the Prometheus instance.
Tbh I would continue to use the ServiceMonitor but I can tune the relabeling configured there to have the `strimzi_io_name` still as `<cluster>-zookeeper` and `<cluster>-kafka` and having the service name available with a new label.

Reply:
That sounds good. The label change is really what changes the compatibility ... apart form that Service Monitor does not matter to me. And since that does not change I do not think we need some special docs about it - just the new users can install the new setup.